### ISSUE TITLE
Command line interface to `NSApp.squirrelAppDelegate.deploy`

### ISSUE THREAD

**Issue by nalzok** [rime/squirrel#1009]
It would be nice to have a command line interface to deploy the changes. For example, I am using https://github.com/nix-community/home-manager to manage the rime configuration files. Instead of manually clicking the menu item "Deploy" each time I switch to a new generation, I want to set [`home.file.<name>.onChange`](https://mynixos.com/home-manager/option/home.file.%3Cname%3E.onChange) to a command/script that programmatically invoke the "Deploy" action. In other words, I'm looking for a command that is similar to `touch ~/.config/ibus/rime/; ibus restart`, but for macOS.

**Comment by LEOYoon-Tsaw** [comment]
It is supported. The command is `--build`

**Comment by nalzok** [comment]
Thanks. Did you mean `/Library/Input\ Methods/Squirrel.app/Contents/MacOS/Squirrel --build`? In addition, I wonder what's the difference between `--build` and `--reload`?

```
$ /Library/Input\ Methods/Squirrel.app/Contents/MacOS/Squirrel --help
Supported arguments:
Perform actions:
  --quit                     quit all Squirrel process
  --reload                   deploy
  --sync                     sync user data
  --build                    build all schemas in current directory
Install Squirrel:
  --install, --register-input-source    register input source
  --enable-input-source [source id...]  input source list optional
  --disable-input-source [source id...] input source list optional
  --select-input-source [source id]     input source optional
```

**Comment by LEOYoon-Tsaw** [comment]
I dont remember on top of my head, you might refer to code in Source/Main.swift to see the details

**Comment by nalzok** [comment]
For the record, I need to run `rm -rf ~/Library/Rime/build` before executing `"/Library/Input Methods/Squirrel.app/Contents/MacOS/Squirrel" --reload`, or the changes to the config file won't be deployed. Perhaps there are some caching issues?

**Comment by ksqsf** [comment]
It could be, but that's unusual. What changes are not being deployed? Is it a dictionary or a config file?

**Comment by nalzok** [comment]
Just config file. In particular, I was updating `~/Library/Rime/default.custom.yaml` and `~/Library/Rime/squirrel.custom.yaml`.

**Comment by ksqsf** [comment]
That's weird. Since in your case `--reload` initiates a full deployment, I doubt that there are some errors in your config. But if that's the case, removing build won't fix it and you will be unable to type anyway. Can you confirm that your changes are correct?

FWIW, librime operates on two levels of deployment: full and quick.

- A full deployment rebuilds all config files (only skipping unchanged dictionaries).

- In the case of a quick check, librime will compare all modification times of all yaml files (under the user dir) to a timestamp `last_build_time` stored in user.yaml, and only when this check fails does librime starts a deployment.

> BTW, is there a way to print out the loaded config so that I can verify the changes have been applied?

Yes. See `build/default.yaml`, etc. They are the final yaml files with all customizations applied.

**Comment by nalzok** [comment]
> Can you confirm that your changes are correct?

Maybe that's the reason! Occasionally I cannot type and have to rollback the configuration and/or reboot the computer. How do I check if my configuration contains an error? I believe the command `--reload` always exits with status code 0 regardless of whether the configuration is correct.

> FWIW, librime operates on two levels of deployment: full and quick.

Thanks, that's good to know. How do I know whether `--reload` does a full or quick deployment?

> I dont remember on top of my head, you might refer to code in Source/Main.swift to see the details

I have checked the source code, but I'm not very familiar with the code base. My educated guess is that `--reload` deals with the config files, whereas `--build` deals with the dictionaries, so they serve completely different purposes. Is that correct?

https://github.com/rime/squirrel/blob/5cde44ae30557f42417a46f2033b3ed68fcdfb44/sources/Main.swift#L37-L39

https://github.com/rime/squirrel/blob/5cde44ae30557f42417a46f2033b3ed68fcdfb44/sources/Main.swift#L70-L78

**Comment by ksqsf** [comment]
> > Can you confirm that your changes are correct?
> 
> Maybe that's the reason! Occasionally I cannot type and have to rollback the configuration and/or reboot the computer. How do I check if my configuration contains an error? I believe the command `--reload` always exits with status code 0 regardless of whether the configuration is correct.

Yes. `--reload` initiates an async deploy request to the current Squirrel instance. You cannot know whether it succeeds or not. The ultimate way to know for sure is to look at the log files: `$TMPDIR/rime.squirrel/*ERROR*`.

If you really need to know whether it succeeds or not immediately, please consider using the `rime_deployer` command bundled in Squirrel.app, e.g.

```
/Library/Input\ Methods/Squirrel.app/Contents/MacOS/rime_deployer --build ~/Library/Rime /Library/Input\ Methods/Squirrel.app/Contents/SharedSupport ~/Library/Rime/build
```

It will exit with non-zero status to indicate any errors. After that you can use `--reload` to update the running Squirrel instance.

> > FWIW, librime operates on two levels of deployment: full and quick.
> 
> Thanks, that's good to know. How do I know whether `--reload` does a full or quick deployment?

It's always a full deployment.

**Comment by nalzok** [comment]
Thank you! The `rime_deployer` command you provided is very helpful. I guess I'll run it, followed by both `--build` and `--deploy`.

There is a problem though: suppose I have the following invalid `~/Library/Rime/default.custom.yaml`, where `hehehe` is a non-existent schema to trigger an error

```yaml
patch:
  schema_list:
    - schema: rime_ice
    - schema: hehehe
```

Now `rime_deployer` reports an error, which expected

```bash
$ /Library/Input\ Methods/Squirrel.app/Contents/MacOS/rime_deployer --build ~/Library/Rime /Library/Input\ Methods/Squirrel.app/Contents/SharedSupport ~/Library/Rime/build
E20241229 15:43:56.943651 0x1f16f0240 deployment_tasks.cc:210] missing input schema: hehehe
```

Now I fix the configuration in `~/Library/Rime/default.custom.yaml`

```yaml
patch:
  schema_list:
    - schema: rime_ice
```

If I run `rime_deployer` again, it gives me the same error, as if the incorrect configuration is cached somewhere

```bash
$ /Library/Input\ Methods/Squirrel.app/Contents/MacOS/rime_deployer --build ~/Library/Rime /Library/Input\ Methods/Squirrel.app/Contents/SharedSupport ~/Library/Rime/build
E20241229 15:44:46.480397 0x1f16f0240 deployment_tasks.cc:210] missing input schema: hehehe
```

There is a reference to `hehehe` in `~/Library/Rime/build/default.yaml`, which is weird: I thought the purpose of `rime_deployer` is to read `~/Library/Rime/default.custom.yaml` and then overwrite `~/Library/Rime/build/default.yaml`. Why would it complain about an error in the old deployment file `~/Library/Rime/build/default.yaml`?

```bash
$ rg hehehe ~/Library/Rime/build
/Users/qys/Library/Rime/build/default.yaml
121:  - schema: hehehe
```

**Comment by nalzok** [comment]
Hi, I noticed another issue: every time I run `/Library/Input\ Methods/Squirrel.app/Contents/MacOS/Squirrel --build`, it generates an `installation.yaml` in the current directory. For example, if I execute it from the home directory, I ends up getting a file at `~/installation.yaml`, which I don't think will be picked up by Squirrel. Am I suppose to run the command in the `~/Library/Rime` directory, so that the output file goes to `~/Library/Rime/installation.yaml`?

BTW, a clarification of the difference between `--build` and `--deploy` is still appreciated, because I don't feel confident about my understanding of your code base.

**Comment by lotem** [comment]
The `Squirrel --reload` command tell the running input method process to do the work. Go for it.

If your target directory is `~/Library/Rime/`, don't use the other commands when the input method is running.

**Comment by nalzok** [comment]
Thanks! I'm getting some conflicted suggestions by different people in this thread, so just to check my understanding is correct:
1. I should run `--deploy` but not `rime_deployer`. This mean I will not be able to check if there is an error in the configuration file immediately. However, I can check `$TMPDIR/rime.squirrel/*ERROR*` after a while to see if there is a problem.
2. I should never run `--build`, even after I have updated/installed a schema in `~/Library/Rime/`. Whenever I run `--deploy`, it will check for changes in the `~/Library/Rime/*.schema.yaml` files, and deploy data to the working directory by invoking `--build` as necessary.

**Comment by ksqsf** [comment]
> I should never run rime_deployer.

Not really. You just cannot run `rime_deployer` while Squirrel is running. You could quit Squirrel before doing that.

**Comment by nalzok** [comment]
Hmmm, I'm afraid the procedure you described isn't working. Basically, I have the following configuration in `~/Library/Rime/default.custom.yaml`:
```
patch:
  menu/page_size: 5
```

However,  no matter how many times I ran `--deploy`, the menu shows 9 items instead of 5. Moreover, clicking on the "Deploy" menu item on the menu bar also doesn't work. See the log file: 
[rime.squirrel.gloaming.qys.log.INFO.20250101-232159.3884.log](https://github.com/user-attachments/files/18287833/rime.squirrel.gloaming.qys.log.INFO.20250101-232159.3884.log) (it's long because I `--deploy`ed multiple times). Apparently Squirrel doesn't realize that there is a `~/Library/Rime/default.custom.yaml` file, since the log does not contain this path.

As a hacky workaround, I did `rm -r ~/Library/Rime/build/` followed by another `--deploy`, and this time the configuration is loaded correctly.

In case it's relevant, one quirky bit of nix-darwin is that it replaces `~/Library/Rime/default.custom.yaml` with a read-only link, but I don't think that makes a difference since Squirrel should be able to detect changes from the file modification time.
```
$ ls -la ~/Library/Rime/default.custom.yaml
lrwxr-xr-x  1 qys  staff  95 Jan  1 23:17 /Users/qys/Library/Rime/default.custom.yaml@ -> /nix/store/aig1ib6l6xziizqs6k13igacqsg3abpn-home-manager-files/Library/Rime/default.custom.yaml
```

**Comment by nalzok** [comment]
> You could quit Squirrel before doing that.

What's the proper way to quit and restart Squirrel? I know I can do `/Library/Input\ Methods/Squirrel.app/Contents/MacOS/Squirrel --quit && /Library/Input\ Methods/Squirrel.app/Contents/MacOS/Squirrel`, but that launches a Squirrel process in the foreground, which is not very convenient in a script. I'm looking for something asynchronous, like `--deploy`.

**Comment by ksqsf** [comment]
> but that launches a Squirrel process in the foreground, which is not very convenient in a script. I'm looking for something asynchronous, like `--deploy`.

Either the traditional unix way or just let macOS restart it automatically.

**Comment by nalzok** [comment]
By "the traditional unix way", did you mean launching it in the background with `/Library/Input\ Methods/Squirrel.app/Contents/MacOS/Squirrel &` and then calling `nohup`?

The "just let macOS restart it automatically" part is what concerns me: how do I know if macOS would restart it when `rime_deployer` is still running, or if I would accidentally launch another Squirrel process in the background when macOS has already restarted it? I mean, it's very hard to ensure Squirrel is not running when macOS can automatically restart it.

**Comment by ksqsf** [comment]
Unfortunately, there is no way to control macOS behavior. I agree the best way is --reload.

**Comment by nalzok** [comment]
Thanks for the help everyone. However, for the record, there is still a bug where I have to `rm -r ~/Library/Rime/build/` before `--deploy` for the new configuration to take effect. See https://github.com/rime/squirrel/issues/1009#issuecomment-2567275768 for details.
